Add scalar conversion using avx instructions for half

CaoE opened the pull request:
### Motivation

Scalar conversion between Half and Float on CPU is more time consuming compared to BFloat16 <-> Float. There is no direct data type conversion instruction for single Half value on CPU, so we add scalar conversion with avx instructions for Half to speed up.

### Testing
Test maxpool, and compared with the results of #98819.
Single socket (28 cores):

shape | fp16 forward / ms | bf16 forward / ms | fp16 backward / ms | bf16 backward / ms | speedup ratio (fp16 forward) | speedup ratio (fp16 backward)
-- | -- | -- | -- | -- | -- | --
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: contig | 5.07165 | 5.418 | 0.5798 | 0.5123 | 1.373694951 | 3.430786
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: CL | 1.37455 | 1.2505 | 8.8336 | 9.7684 | 1.373635008 | 4.132924
size: (32, 16, 200, 200), kernel: 3,   stride: 1, mem_format: contig | 28.72 | 30.7069 | 3.813 | 3.75 | 1.31977124 | 2.783006
size: (32, 16, 200, 200), kernel: 3,   stride: 1, mem_format: CL | 4.5783 | 4.703 | 4.703 | 5.1 | 1.028980189 | 3.1293
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: contig | 13.896 | 14.8138 | 1.6635 | 1.6274 | 1.298704663 | 2.982699
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: CL | 2.11291 | 2.1158 | 2.26778 | 2.272 | 0.951105348 | 3.179012
size: (4, 19, 10, 16, 16), kernel: 3,   stride: 1, mem_format: contig | 0.4204 | 0.3843 | 0.0649 | 0.0633 | 2.102711703 | 1.779492
size: (4, 19, 10, 16, 16), kernel: 3,   stride: 1, mem_format: CL3d | 0.1134 | 0.11 | 0.1476 | 0.143 | 2.23042328 | 3.612398



Single core:

shape | fp16 forward / ms | bf16 forward / ms | fp16 backward / ms | bf16 backward / ms | speedup ratio (fp16 forward) | speedup ratio (fp16 backward)
-- | -- | -- | -- | -- | -- | --
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: contig | 124.413 | 114.44 | 10.553 | 11.2486 | 1.31395433 | 3.923844
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: CL | 28.99 | 28.0781 | 9.5092 | 10.9258 | 1.324296999 | 3.888377
size: (32, 16, 200, 200), kernel: 3,   stride: 1, mem_format: contig | 640.8276 | 591.964 | 59.18776 | 60.854 | 1.334956391 | 3.704458
size: (32, 16, 200, 200), kernel: 3,   stride: 1, mem_format: CL | 88.57 | 90.214 | 54.358 | 59.205 | 1.031258214 | 3.75285
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: contig | 318.6197 | 285.155 | 28.4999 | 29.4387 | 1.315298144 | 3.759747
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: CL | 31.3981 | 34.0544 | 25.6557 | 28.7811 | 1.068505738 | 3.841587
size: (4, 19, 10, 16, 16), kernel: 3,   stride: 1, mem_format: contig | 8.87882 | 8.207 | 0.386056 | 0.3939 | 1.567866 | 3.50387
size: (4, 19, 10, 16, 16), kernel: 3,   stride: 1, mem_format: CL3d | 2.4167 | 2.38295 | 0.3769 | 0.4066 | 1.39402491 | 3.30061

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

CaoE committed:
add scalar conversion using avx instructions for half

ghstack-source-id: f15119220f620f2142f4457e4f5dae2e136e14b7
Pull Request resolved: https://github.com/pytorch/pytorch/pull/101378

CaoE commented:
Moved https://github.com/pytorch/pytorch/pull/101378 from stack to here.

mingfeima reviewed (approved):
LGTM now

CaoE commented:
@ngimel Could please review this PR ? Thank you.

CaoE commented:
@cpuhrsch Could you please review this PR ? Thanks.

cpuhrsch commented on aten/src/ATen/cpu/vec/vec_half.h:
```diff
@@ -0,0 +1,40 @@
+inline namespace CPU_CAPABILITY {
+
+#if defined(CPU_CAPABILITY_AVX2) || defined(CPU_CAPABILITY_AVX512)
+inline uint16_t float2half_scalar(float val) {
```
Why introduce these new functions that are only defined under both of these options?

It looks like they're only used once.

CaoE replied:
There is no direct data type conversion instruction for single Half value on CPU, so I add scalar conversion with AVX instructions for Half to speed up when AVX2 or AVX512 is supported on the platform.

cpuhrsch commented on c10/util/Half-inl.h:
```diff
@@ -48,9 +56,13 @@ inline C10_HOST_DEVICE Half::operator float() const {
   return float(c10::bit_cast<sycl::half>(x));
+#else
+#if defined(CPU_CAPABILITY_AVX2) || defined(CPU_CAPABILITY_AVX512)
+  return at::vec::half2float_scalar(x);
```
Should we just inline the code here?

cpuhrsch replied:
Alternatively we can unify code that works for all CPU types in the half2float, float2half conversion functions.

CaoE replied:
> Should we just inline the code here?

I can not inline `at::vec::half2float_scalar` or `at::vec::float2half_scalar` in `c10/util/Half-inl.h` since PyTorch build systems does not add compilation flags related to AVX or AVX512 for `c10/util/Half-inl.h` during compilation. It is unable to compile AVX related instructions in `c10/util/Half-inl.h`. That's why I chose to add these two functions under aten/src/Aten/cpu.

cpuhrsch reviewed (changes requested):
Are we sure that CI covers these changes? Can you raise an exception within the conversion and see which tests and which environment fails? I want to make sure we have a CPU with the required capability and a test that exercises this.

CaoE committed:
add a test case for half conversion

CaoE commented:
@cpuhrsch I added a c++ test case for half conversion. 

> Can you raise an exception within the conversion and see which tests and which environment fails?

Half conversion would not raise an exception since if the cpu does not support avx2 or avx512 it will fallback to `fp16_ieee_from_fp32_value` and `detail::fp16_ieee_to_fp32_value`.

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
@pytorchbot merge

CaoE commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m="This is breaking internal tests @cpuhrsch can share more context and help with a follow up" -c=nosignal

cpuhrsch commented:
@CaoE - This caused issues on Mac: `stderr: clang: error: unable to execute command: Illegal instruction: 4`

CaoE commented:
@cpuhrsch May I know if the error occurred during the build phase or what test case caused it ? Could you please provide a reproducer to help us reproduce this issue ? Thank you.

CaoE commented:
@cpuhrsch I didn't reproduce it on the Mac machine in aws (mac1.metal) with Apple clang version 13.0.0.

cpuhrsch commented:
@CaoE - I can do some more digging later. At the moment I don't have time to follow up and resolve this internally. If you can reproduce the issue in the OSS, then that'd help. Otherwise this will have to wait about two weeks.

malfet commented:
@CaoE this happened with xcode_14.3.1, which I happened to have locally let's see if I can easily repro the failure

cpuhrsch commented:
@CaoE - Do you have access to an Intel Mac for this? The error seems to happen on Intel Mac.

cpuhrsch commented:
@CaoE - Great! Thanks for resolving this. We can import the PR first once all OSS jobs ran green to get internal test signal and then merge if all worked out.

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
This looks like an unrelated failure. @CaoE could you please rebase and we'll try again?

CaoE commented:
@cpuhrsch Thanks for your help! I rebased and also removed apple check by using `_cvtss_sh` and `_cvtsh_ss` instead, not sure if it can pass internal building on mac and tests.

CaoE commented:
`_cvtss_sh` and `_cvtsh_ss` will still meet same issue with xcode_14.3.1 on mac. Add apple check back.

CaoE commented:
@pytorchbot merge
